### PR TITLE
bpo-40501: Replace ctypes code in uuid with native module

### PR DESCRIPTION
**Pull request by zooba**
https://bugs.python.org/issue40501

### CODE REVIEW

File: Modules/_uuidmodule.c [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Python UUID module that wraps libuuid -
+ * Python UUID module that wraps libuuid or rpcrt4.dll
```
Comment (vstinner) [inline Modules/_uuidmodule.c]:
```suggestion
 * Python UUID module that wraps libuuid or Windows rpcrt4.dll.
```

File: Lib/uuid.py [hunk 2]
```diff
@@ -555,68 +555,15 @@ def _netstat_getnode():
...
-                    mac = int(value.replace('-', ''), 16)
-                    if _is_universal(mac):
-                        return mac
-                    first_local_mac = first_local_mac or mac
-    return first_local_mac or None
+    """[DEPRECATED] Get the hardware address on Windows."""
+    # bpo-40501: UuidCreateSequential() is now the only supported approach
+    return _windll_getnode()
 
 def _netbios_getnode():
```
Comment (vstinner) [inline Lib/uuid.py:562]: Why not simply removing _ipconfig_getnode and _netbios_getnode? They are private functions, there is no backward compatibility warranty.

File: Modules/_uuidmodule.c [hunk 3]
```diff
@@ -31,19 +37,51 @@ py_uuid_generate_time_safe(PyObject *Py_UNUSED(context),
...
 static int
 uuid_exec(PyObject *module) {
     assert(sizeof(uuid_t) == 16);
-#ifdef HAVE_UUID_GENERATE_TIME_SAFE
+#if defined(HAVE_UUID_GENERATE_TIME_SAFE)
     int has_uuid_generate_time_safe = 1;
 #else
+    /* including on Windows */
     int has_uuid_generate_time_safe = 0;
 #endif
```
Comment (vstinner) [inline Modules/_uuidmodule.c]: Maybe be even more explicit, something like:

```
#ifdef MS_WINDOWS
     int has_uuid_generate_time_safe = 0; 
#elif defined(HAVE_UUID_GENERATE_TIME_SAFE)
     int has_uuid_generate_time_safe = 1; 
#else
     int has_uuid_generate_time_safe = 0; 
#endif
```

**Review by vstinner — COMMENTED** [review]
Can you please try to leave the Lib/uuid.py ctypes code for libuuid unchanged in this PR? I would prefer to remove it in a separated PR.

File: Lib/uuid.py (hunk 2 above)
Comment (zooba) [reply Lib/uuid.py:562]: Last time I was involved in one of those people still broke. I'd rather leave them there as long as they aren't hurting anyone.

**Comment by zooba** [comment]
> I would prefer to remove it in a separated PR.

Why? It's dead code, and you're just as likely to come back and ask me to not leave dead code in there.

**Comment by vstinner** [comment]
> Why? It's dead code, and you're just as likely to come back and ask me to not leave dead code in there.

I'm not confident yet that it's safe to remove it. For example, nobody checked FreeBSD so far: https://bugs.python.org/issue40501#msg368068

Leave it for now would allow to at least merge the Windows part.

**Comment by zooba** [comment]
I've checked BSD now :)

Apparently there are new leaks in subinterpreter tests, which I'm pretty sure weren't me, but as they're only showing up in my PR I guess I'll take a look.

Nice to see my installer check running as well!

**Comment by zooba** [comment]
Refleaks are confirmed as coming from a different PR, so I'm merging this as it passes on the rest of the fleet.

**Pull request merged**